clients/src/main/java/org/apache/kafka/common/record/EosLogRecord.java

```diff
@@ -25,215 +25,154 @@
-                             byte attributes,
-                             long timestamp,
-                             ByteBuffer key,
-                             ByteBuffer value) throws IOException {
-        int recordSize = sizeOf(key, value);
-
-        if (key == null)
-            attributes |= NULL_KEY_MASK;
-        else
-            attributes &= ~NULL_KEY_MASK;
+    public static long writeTo(DataOutputStream out,
+                               int offsetDelta,
+                               long timestamp,
+                               ByteBuffer key,
+                               ByteBuffer value) throws IOException {
+        int sizeInBytes = sizeOfBodyInBytes(offsetDelta, timestamp, key, value);
+        Utils.writeUnsignedVarInt(sizeInBytes, out);
 
-        if (value == null)
-            attributes |= NULL_VALUE_MASK;
-        else
-            attributes &= ~NULL_VALUE_MASK;
+        byte attributes = computeAttributes(key, value);
+        out.write(attributes);
 
-        out.writeInt(recordSize);
-        out.writeLong(offset);
-        out.writeByte(attributes);
-        out.writeLong(timestamp);
+        Utils.writeVarLong(timestamp, out);
+        Utils.writeUnsignedVarInt(offsetDelta, out);
 
         if (key != null) {
-            attributes &= ~NULL_KEY_MASK;
-            int size = key.remaining();
-            out.writeInt(size);
-            out.write(key.array(), key.arrayOffset(), size);
+            int keySize = key.remaining();
+            Utils.writeUnsignedVarInt(keySize, out);
+            out.write(key.array(), key.arrayOffset(), keySize);
         }
 
         if (value != null)
             out.write(value.array(), value.arrayOffset(), value.remaining());
 
-        long crc = computeChecksum(recordSize, offset, attributes, timestamp, key, value);
-        out.writeInt((int) (crc & 0xffffffffL));
-        return crc;
+        return computeChecksum(timestamp, key, value);
+    }
+
+    public static long writeTo(ByteBuffer out,
+                               int offsetDelta,
+                               long timestamp,
+                               ByteBuffer key,
+                               ByteBuffer value) {
+        try {
+            return writeTo(new ByteBufferOutputStream(out), offsetDelta, timestamp, key, value);
+        } catch (IOException e) {
+            // cannot actually be raised by ByteBufferOutputStream
+            throw new RuntimeException(e);
+        }
     }
 
     /**
      * Compute the checksum of the record from the attributes, key and value payloads
      */
-    public static long computeChecksum(int recordSize,
-                                       long offset,
-                                       byte attributes,
-                                       long timestamp,
-                                       ByteBuffer key,
-                                       ByteBuffer value) {
+    private static long computeChecksum(long timestamp,
```

Comment:
I don't think it's necessary. The main thing we are trying to achieve is consistency between what the producer sees when writing and what the consumer sees when reading. It was necessary before when we were using the record checksum to validate the full message, but now we depend on the entry-level checksum for that.